File: oldnew/freeworld/world/World.java
/*
 * MIT License
 *
 * Copyright (c) 2021 OverRun Organization
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.overrun.freeworld.world;

import org.overrun.freeworld.block.Block;
import org.overrun.freeworld.util.Registry;
import org.overrun.freeworld.world.phys.AABB;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.EOFException;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

import static org.overrun.freeworld.FreeWorld.logger;
import static org.overrun.freeworld.block.Blocks.*;

/**
 * @author squid233
 * @since 2021/05/04
 */
public final class World {
    private static final int BLOCK_UPDATE_INTERVAL = 400;
    public final int width;
    public final int height;
    public final int depth;
    private final Block[] blocks;
    private final int[] lightDepths;
    private final List<WorldListener> listeners = new ArrayList<>();
    private final List<AABB> aabbs = new ArrayList<>();
    private final Random random = new Random();
    private int unprocessed;

    public World(int width, int height, int depth) {
        this.width = width;
        this.height = height;
        this.depth = depth;
        blocks = new Block[width * height * depth];
        lightDepths = new int[width * depth];
        Arrays.fill(blocks, AIR);
        boolean loaded = load();
        if (!loaded) {
            generateMap();
        }
        calcLightDepths(0, 0, width, depth);
    }

    private void generateMap() {
        for (int x = 0; x < width; x++) {
            for (int z = 0; z < depth; z++) {
                for (int y = 0; y < 5; y++) {
                    setBlock(COBBLESTONE, x, y, z);
                    setBlock(DIRT, x, y + 5, z);
                }
                setBlock(GRASS_BLOCK, x, 10, z);
            }
        }
    }

    public void save() {
        try (var fos = new FileOutputStream("world.dat");
             var gos = new GZIPOutputStream(fos);
             var dos = new DataOutputStream(gos)) {
            //noinspection ForLoopReplaceableByForEach
            for (int i = 0; i < blocks.length; i++) {
                dos.writeInt(blocks[i].getRawId());
            }
        } catch (IOException e) {
            logger.catching(e);
        }
    }

    public boolean load() {
        try (var fis = new FileInputStream("world.dat");
             var gis = new GZIPInputStream(fis);
             var dis = new DataInputStream(gis)) {
            try {
                //noinspection InfiniteLoopStatement
                for (int i = 0, read; ; i++) {
                    read = dis.readInt();
                    blocks[i] = Registry.BLOCK.get(read);
                }
            } catch (EOFException ignored) {
            }
            calcLightDepths(0, 0, width, depth);
            for (WorldListener listener : listeners) {
                listener.allChanged();
            }
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    public void calcLightDepths(int x0, int z0, int x1, int z1) {
        for (int x = x0; x < x1; x++) {
            for (int z = z0; z < z1; z++) {
                int oldDepth = lightDepths[x + z * width];
                int y = depth - 1;
                while (y > 0 && isOpacity(x, y, z)) {
                    --y;
                }
                lightDepths[x + z * width] = y;
                if (oldDepth != y) {
                    int yl0 = Math.min(oldDepth, y);
                    int yl1 = Math.max(oldDepth, y);
                    for (WorldListener listener : listeners) {
                        listener.lightColumnChanged(x, y, yl0, yl1);
                    }
                }
            }
        }
    }

    public void tick() {
        unprocessed += width * height * depth;
        int ticks = unprocessed / BLOCK_UPDATE_INTERVAL;
        unprocessed -= ticks * BLOCK_UPDATE_INTERVAL;
        for (int i = 0; i < ticks; i++) {
            int x = random.nextInt(width);
            int y = random.nextInt(height);
            int z = random.nextInt(depth);
            getBlock(x, y, z).randomTick(this, x, y, z, random);
        }
    }

    public void addListener(WorldListener listener) {
        listeners.add(listener);
    }

    public List<AABB> getCollision(AABB aabb) {
        aabbs.clear();
        int x0 = Math.max((int) aabb.x0, 0);
        int y0 = Math.max((int) aabb.y0, 0);
        int z0 = Math.max((int) aabb.z0, 0);
        int x1 = Math.min((int) (aabb.x1 + 1), width);
        int y1 = Math.min((int) (aabb.y1 + 1), height);
        int z1 = Math.min((int) (aabb.z1 + 1), depth);
        for (int y = y0; y < y1; y++) {
            for (int x = x0; x < x1; x++) {
                for (int z = z0; z < z1; z++) {
                    AABB aabb1 = getBlock(x, y, z).getCollision().move(x, y, z);
                    if (aabb1 != AABB.empty()) {
                        aabbs.add(aabb1);
                    }
                }
            }
        }
        return aabbs;
    }

    public boolean isOpacity(int x, int y, int z) {
        return getBlock(x, y, z).isOpacity();
    }

    public boolean isLit(int x, int y, int z) {
        return !inWorld(x, y, z) || y >= lightDepths[x + z * width];
    }

    public boolean isSolidBlock(int x, int y, int z) {
        return getBlock(x, y, z).isSolid();
    }

    public boolean inWorld(int x, int y, int z) {
        return x >= 0 && x < width &&
                y >= 0 && y < height &&
                z >= 0 && z < depth;
    }

    public int getIndex(int x, int y, int z) {
        return (y * depth + z) * width + x;
    }

    public void setBlock(Block block, int x, int y, int z, boolean update) {
        if (inWorld(x, y, z)) {
            blocks[getIndex(x, y, z)] = block == null ? AIR : block;
            calcLightDepths(x, z, 1, 1);
            if (update) {
                for (WorldListener listener : listeners) {
                    listener.blockChanged(x, y, z);
                }
            }
        }
    }

    public void setBlock(Block block, int x, int y, int z) {
        setBlock(block, x, y, z, true);
    }

    public Block getBlock(int x, int y, int z) {
        if (inWorld(x, y, z)) {
            return blocks[getIndex(x, y, z)];
        }
        return AIR;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int getDepth() {
        return depth;
    }
}

File: src/main/kotlin/io/github/overrun/freeworld/world/World.kt
/*
 * MIT License
 *
 * Copyright (c) 2021 OverRun Organization
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package io.github.overrun.freeworld.world

import io.github.overrun.freeworld.block.Block
import io.github.overrun.freeworld.block.Blocks.*
import io.github.overrun.freeworld.client.GlProgram
import io.github.overrun.freeworld.client.Transformation
import io.github.overrun.freeworld.entity.player.Player
import org.joml.Matrix4f
import org.lwjgl.opengl.GL11.*

/**
 * This is [world][World].
 * This is not [free][io.github.overrun.freeworld.FreeWorld].
 * @author squid233
 * @since 2021/03/24
 */
class World(
    val width: Int,
    val height: Int,
    val depth: Int
) {
    private val blocks = Array(width * height * depth) { air }

    init {
        val layer = 2
        for (x in 0 until width) {
            for (z in 0 until depth) {
                for (y in 0 until layer) {
                    setBlock(x, y, z, dirt)
                }
                setBlock(x, layer, z, grassBlock)
            }
        }
    }

    fun render(program: GlProgram, transformation: Transformation, viewMatrix: Matrix4f) {
        var i = 0
        for (x in 0 until width) {
            for (z in 0 until depth) {
                for (y in 0 until height) {
                    val block = getBlock(x, y, z)
                    block.x = x
                    block.y = y
                    block.z = z
                    if (block != air && (
                                getBlock(x - 1, y, z) == air
                                        || getBlock(x + 1, y, z) == air
                                        || getBlock(x, y - 1, z) == air
                                        || getBlock(x, y + 1, z) == air
                                        || getBlock(x, y, z - 1) == air
                                        || getBlock(x, y, z + 1) == air
                                )
                    ) {
                        program.setUniform(
                            "modelViewMatrix",
                            transformation.getModelViewMatrix(block, viewMatrix)
                        )
                        var result = 0
                        if (getBlock(x - 1, y, z) == air)
                            result += FACE_LEFT + FACE_OVERLAY_LEFT
                        if (getBlock(x + 1, y, z) == air)
                            result += FACE_RIGHT + FACE_OVERLAY_RIGHT
                        if (getBlock(x, y - 1, z) == air)
                            result += FACE_BOTTOM
                        if (getBlock(x, y + 1, z) == air)
                            result += FACE_TOP
                        if (getBlock(x, y, z - 1) == air)
                            result += FACE_BACK + FACE_OVERLAY_BACK
                        if (getBlock(x, y, z + 1) == air)
                            result += FACE_FRONT + FACE_OVERLAY_FRONT
                        block.render(result)
                    }
                    ++i
                }
            }
        }
    }

    fun pick(program: GlProgram, transformation: Transformation, viewMatrix: Matrix4f) {
        glInitNames()
        glPushName(0)
        glPushName(0)
        val boxRadius = 6
        val pX0 = (Player.x - boxRadius).toInt()
        val pY0 = (Player.y - boxRadius).toInt()
        val pZ0 = (Player.z - boxRadius).toInt()
        val pX1 = (Player.x + boxRadius + 1).toInt()
        val pY1 = (Player.y + boxRadius + 1).toInt()
        val pZ1 = (Player.z + boxRadius + 1).toInt()
        var i = 0
        for (x in pX0 until pX1) {
            glLoadName(x)
            glPushName(0)
            for (y in pY0 until pY1) {
                glLoadName(y)
                glPushName(0)
                for (z in pZ0 until pZ1) {
                    glLoadName(z)
                    glPushName(0)
                    val block = getBlock(x, y, z)
                    block.x = x
                    block.y = y
                    block.z = z
                    if (block != air && (
                                getBlock(x - 1, y, z) == air
                                        || getBlock(x + 1, y, z) == air
                                        || getBlock(x, y - 1, z) == air
                                        || getBlock(x, y + 1, z) == air
                                        || getBlock(x, y, z - 1) == air
                                        || getBlock(x, y, z + 1) == air
                                )
                    ) {
                        program.setUniform(
                            "modelViewMatrix",
                            transformation.getModelViewMatrix(block, viewMatrix)
                        )
                        glLoadName(FACE_FRONT)
                        block.render(FACE_FRONT)
                        glLoadName(FACE_RIGHT)
                        block.render(FACE_RIGHT)
                        glLoadName(FACE_TOP)
                        block.render(FACE_TOP)
                        glLoadName(FACE_LEFT)
                        block.render(FACE_LEFT)
                        glLoadName(FACE_BACK)
                        block.render(FACE_BACK)
                        glLoadName(FACE_BOTTOM)
                        block.render(FACE_BOTTOM)
                    }
                    glPopName()
                    ++i
                }
                glPopName()
            }
            glPopName()
        }
        glPopName()
        glPopName()
    }

    fun inBound(x: Int, y: Int, z: Int) =
        x in 0 until width
                && y in 0 until height
                && z in 0 until depth

    fun getIndex(x: Int, y: Int, z: Int) =
        //(x % width) + (y * width) + (z * width * height)
        (x) + (y * width * depth) + (z * width)

    fun getBlock(x: Int, y: Int, z: Int) =
        if (inBound(x, y, z)) blocks[getIndex(x, y, z)]
        else air

    @JvmOverloads
    fun setBlock(x: Int, y: Int, z: Int, block: Block, force: Boolean = false) {
        if (!inBound(x, y, z)) return
        if (!force && getBlock(x, y, z) != air) return
        blocks[getIndex(x, y, z)] = block
    }
}